feat: Add Super.json reference (#37)
superfaceai/docs: docs/reference/superjson.mdx

---
slug: /reference/superjson
---

# Super.json Reference

> Super.json is the main Superface configuration file. It contains the configuration about profiles, maps and providers used.

Reviewer: Preferably avoid capitalization (same for title), the file is called `super.json` and most apps will likely run on case-sensitive FS. :wink: 

```suggestion
`super.json` is the main Superface configuration file. It specifies which profiles, maps, and providers are used.
```

Maybe also mention where the file needs to be located (in `superface` directory).

## Anatomy

> At the top level, super.json file consists of two sections: [profiles](#profiles) and [providers](#providers). the main purpose profile section is to set up profiles which are to be used, their versions and/or file locations, and their default values. The purpose of provider section is to set up providers that are to be used and their credentials.

Reviewer 2: ```suggestion
At the top level, super.json file consists of two sections: [profiles](#profiles) and [providers](#providers). Profiles section contains list of installed profiles, their versions and/or file locations, and default values. Providers section contains list of installed providers and their credentials.
```
I tried to write it in simple English. Not sure about the term `used`. I would use `installed`.

Reviewer: Working further from Jan's suggestion to add articles:
```suggestion
At the top level, the `super.json` file consists of two sections: [profiles](#profiles) and [providers](#providers). The profiles section contains a list of installed profiles, their versions or file locations, and default values. The providers section contains a list of installed providers and their credentials.
```

> ```json

Reviewer: ````suggestion
```json title="superface/super.json" 
````

Consider adding this title to all code examples.

{
  "profiles": {

>     // ...

Reviewer: In general it would be helpful to mention at least the object name in these comments instead of ellipses.

  },
  "providers": {
   // ...
  }
}
```

## Profiles

The profiles section consists of a record of all installed profiles. The key is the name of the profile and value is either semantic version string, a file URI or a [ProfileSettings](#profilesettings) object.

```json
{
  "profiles": {
    "myprofile1": "1.0.0",
    "myprofile2": "file:path/to/profile.supr",
    "myprofile3": {
      // ...
    }
  }
}
```

### ProfileSettings

> The ProfileSettings object contains either `file` or `version` property, analogous to the shortcut strings above and following optional properties:

Reviewer: I miss an explanation of `file` and `version` properties and their shortcut variants. Especially for file: is it relative path or absolute path or what? This is not self explanatory.

- `priority`: an array of strings containing provider names in desired order for failover
- `providers`: an array of [ProfileProviderSettings](#profileprovidersettings) objects or URI strings, used for configuring providers for the profile. Not to be confused with top-level Providers record

> - `defaults`: a record of defaults for the usecase, key is usecase name, value is [UsecaseDefaults](#usecasedefaults) object

Reviewer: So "usecase" is not correct spelling in English. I am trying to keep "use-case" in other parts of the docs.

```json
{
  "version": "1.0.0",

>   "priority": ["provider1", "provider2"],

Reviewer: Provider names **must not** contain numbers, this is invalid.

Author: since when? and where is it specified? profile names use DocumentIdentifier afaik (/[a-z][a-z0-9_-]*/)

Reviewer: See the schema https://superface.ai/docs/comlink/reference/provider#json-schema
Not sure when this was added (back in August it was discussed as given). The rationale is to prevent typo attacks (eg. `google` vs `goog1e`).

Author: and why is it only in this one document (where does the schema even come from?) and not in other documents or in code?

Reviewer: I guess it was missing in the spec and wasn't reflected in the authoring guide. I will add an issue to fix it, the authoring guide needs a major revision anyway.

  "defaults": {

>     "myusecase": {

Reviewer: AFAIK use-cases are always CamelCase?

```suggestion
    "MyUseCase": {
```

      // ...
    }
  },
  "providers": {
    "provider1": {
      // ...
    },

>     "provider2": "file:path/to/provider.json"

Reviewer: Huh, so I can specify the provider locally just for a single profile?

Author: Yes

  }
}
```

### UsecaseDefaults

> UsecaseDefaults is a record used to configure default values for a usecase, independent of provider. The key is usecase name and the value is an object with following optional properties.

Reviewer: Record or object? Also 

```suggestion
UsecaseDefaults specifies default values for a use-case for all providers. The key is use-case's name and the value is an object with following optional properties:
```

- `input`: an object providing default values according to usecase's input

> - `providerFailover`: a boolean specific whether provider failover should be allowed for the usecase

Reviewer: What are the implications?

```json
{
  "myusecase": {
    "input": {
      "someInputProperty": 42
    },
    "providerFailover": true
  }
}
```

### ProfileProviderSettings

> ProfileProviderSettings object is used for configuring specific providers for a profile. It can point to either file or specific variant and revision. It contains following optional property:

Reviewer: In general and in line with what @janhalama suggested, avoid passive voice.
Also I am confused here. In one part you point to a provider, in other you specify mapVariant and mapRevision. So take my suggestion here with a grain of salt.

```suggestion
ProfileProviderSettings object defines profile-specific configuration of providers. The provider definition can be either a file path (in `file` property), or a specific variant and revision (with `mapVariant` and `mapRevision`). It can also contain the optional property `defaults`:
```

- `defaults`: a record of defaults per usecase, key is usecase name and value is a [ProfileProviderDefaults](#profileproviderdefaults) object

> ```json

Reviewer: Provide explaining titles for these examples. Also it's hard to navigate how deeply nested we are in the super.json file, so preferably provide the whole context and use lines highlighting to pinpoint the relevant part.

```

### RetryPolicy

RetryPolicy is used for setting up retry policy for a provider and usecase. There are currently two types of retry policy, `none` or `circuit-breaker`. When using `none`, no retry is attempted. When using `circuit-breaker`, several attemps are made with backoff. Following properties are used to configure the policy (all optional except kind). When using default values, `none` or `circuit-breaker` string can be used as shorthand instead of the object.

- `kind`: `none` or `circuit-breaker`
- `maxContiguousRetries`: maximum number of retry attempts, positive integer, default 5
- `requestTimeout`: timeout after which a request fails, positive integer, milliseconds, default 30000
- `backoff`: either string `exponential` or an object with following properties:
    - `kind`: string `exponential`
    - `start`: the initial delay time, optional, positive integer, milliseconds, default 2000
    - `factor`: the factor by which is the delay time multiplied, optional, positive integer, default 2

```json
{
  "kind": "circuit-breaker",
  "maxContiguousRetries": 3,
  "requestTimeout": 60000,
  "backoff": {
    "kind": "exponential",
    "start": 1000,
    "factor": 3
  }
}

> ```

Reviewer: Add shorthand example.

## Providers

The providers section consists of a record of all used providers. The key is the name of the provider and the value is either a file URI or a [ProviderSettings](#providersettings) object.

```json
{
  "providers": {
    "provider1": "file:path/to/provider.json",
    "provider2": {
      // ...
    }
  }
}
```

### ProviderSettings

The ProviderSettings object has following optional properties:

- `file`: a file URI string
- `security`: an array of [SecurityValues](#securityvalues) objects

```json
{
  "file": "file:path/to/provider.json",
  "security": [
      {
        // ...
      }
  ]
}
```

> ### SecurityValues

Reviewer: I don't see any explanation for using `$` for injecting env variables. Can it be used anywhere or only for security values?

SecurityValues is an union of several types of security configuration objects: API Key, Basic Auth, Bearer Token or Digest. All of them have a property in common:

- `id`: identifier of the security scheme, referenced in map

#### API Key 
API key uses provided API key in a header, body, path or query string (specified in provider.json)

- `apikey`: string containing API key

#### Basic Auth
Basic Auth uses provided username and password and sends them base64 encoded in Authorization header as per [RFC 7617](https://datatracker.ietf.org/doc/html/rfc7617)

- `username`: string containing user name
- `password`: string containing password

#### Bearer Token
Bearer Token uses provided token and sends it in Authorization header as per [RFC 6750](https://datatracker.ietf.org/doc/html/rfc6750)

- `token`: string containing bearer token

#### Digest
Digest uses provided token and sends it in Authorization header as per [RFC 2617](https://datatracker.ietf.org/doc/html/rfc2617).

> *This is not implemented yet.*

Reviewer: If it's not implemented, comment it out with TODO

- `digest`: string containing digest

```json
[
  {
    "id": "first",
    "apikey": "my-api-key"
  },
  {
    "id": "second",
    "username": "myname",
    "password": "mypass"
  },
  {
    "id": "third",
    "token": "my-bearer-token"
  },
  {
    "id": "fourth",
    "digest": "my-digest"
  }
]
```

> ## Example

Reviewer: Consider moving the example at the start of the document, optionally with comments annotating the specific parts. This would be probably more helpful than the current "Anatomy" example.
